ls1intum/Artemis
# [Feature] Store static code analysis configuration

### Checklist

### Motivation and Context



![configurationGradingDataModel](https://user-images.githubusercontent.com/16407766/93869091-a8df7980-fccb-11ea-8d16-084c75419c14.png)

Instructors should be able to tailor static code analysis settings to their specific needs. Within Artemis, instructors will be able to configure static code analysis categories. Checks performed by static code analysis tools are already grouped into categories by the respective tool. As the number of categories is too high and multiple categories of different tools have a semantic overlap, we define a hard-coded mapping to Artemis specific categories. These default categories are generated for each created programming exercise with static code analysis enabled.
For the grading of static code analysis feedback a score penalty system will be implemented. Instructors are able to choose whether a specific category is inactive, the corresponding feedback is only shown to the students or also graded. Instructors are able to define the score penalty for each issue found within a category. At the same time the maximum penalty can be limited for each category or for the whole exercise.

### Description

User interface for updating the categories is implemented in https://github.com/ls1intum/Artemis/pull/2037.

- [x] Database changes for the configuration and grading concept
- [x] On application start-up, read the static code analysis default configurations for each programming language and make them available as a bean
- [x] Create default static code analysis categories when a programming exercise with static code analysis enabled is created
- [x] Also import static code analysis categories when a programming exercise is imported
- [x] Fix import of test cases (bonus multiplier and bonus points were not set)
- [x] Create Resource, Service, Repository for the creation, reading and updating of static code analysis categories. Stand-alone deletion is currently not supported. The categories will be deleted when the programming exercise is removed.
- [x] Improve validation of programming exercises in `ProgrammingExerciseResource`
- [x] Improve test coverage significantly

The actual grading strategy will be implemented in another PR.

### Steps for Testing


Functionality can only be tested after https://github.com/ls1intum/Artemis/pull/2037 has been merged. Just ensure that the database changes don't break anything

1. Create a programming exercise with and without static code analysis enabled
2. Import a programming exercise with and without static code analysis enabled
2. Participate in those exercises and make sure everything works as usual
3. Delete the exercises

### Test Coverage

![PEResource](https://user-images.githubusercontent.com/16407766/93469224-80810500-f8f0-11ea-961d-676689a125c7.png)
![ScaResourceCoverage](https://user-images.githubusercontent.com/16407766/93469227-81199b80-f8f0-11ea-8e10-e3370e510b31.png)
![ServiceCoverage](https://user-images.githubusercontent.com/16407766/93469229-81b23200-f8f0-11ea-8fed-c3ea9ce5aebe.PNG)

### Screenshots


![image](https://user-images.githubusercontent.com/16407766/94629733-e774c580-02c3-11eb-8748-6138e2ecd413.png)
